File: Behavioral/Observer/cpp/test_observer.cpp
#include "observer.hpp"
#include <gtest/gtest.h>
#include <memory>
#include <sstream>

// Custom stream to capture output for assertions
std::ostringstream outputStream;

/**
 * @brief A TestObserver that overrides update to write to the outputStream.
 */
class TestObserver : public ConcreteObserver {
public:
	TestObserver(const std::string &name, std::shared_ptr<Subject> subject)
		: ConcreteObserver(name, subject) {}

	void update() override {
		outputStream << "Observer " << getName() << " received update." << std::endl;
	}
};

// Helper function to reset the output stream
void resetOutputStream() {
	outputStream.str("");
	outputStream.clear();
}

/**
 * @brief Test attaching and notifying a single observer.
 */
TEST(ObserverPatternTest, AttachAndNotifySingleObserver) {
	auto subject = std::make_shared<ConcreteSubject>();
	auto observer = std::make_shared<TestObserver>("Observer1", subject);

	// Attach the observer to the subject
	subject->attach(observer);

	// Trigger an update
	resetOutputStream();
	subject->setState("State1");

	// Assert the correct output
	EXPECT_EQ(outputStream.str(), "Observer Observer1 received update.\n");
}

/**
 * @brief Test attaching multiple observers and verifying notifications.
 */
TEST(ObserverPatternTest, AttachAndNotifyMultipleObservers) {
	auto subject = std::make_shared<ConcreteSubject>();
	auto observer1 = std::make_shared<TestObserver>("Observer1", subject);
	auto observer2 = std::make_shared<TestObserver>("Observer2", subject);

	// Attach the observers to the subject
	subject->attach(observer1);
	subject->attach(observer2);

	// Trigger an update
	resetOutputStream();
	subject->setState("State2");

	// Assert the correct output
	EXPECT_EQ(outputStream.str(),
			  "Observer Observer1 received update.\n"
			  "Observer Observer2 received update.\n");
}

/**
 * @brief Test detaching an observer and ensuring it no longer receives updates.
 */
TEST(ObserverPatternTest, DetachObserver) {
	auto subject = std::make_shared<ConcreteSubject>();
	auto observer1 = std::make_shared<TestObserver>("Observer1", subject);
	auto observer2 = std::make_shared<TestObserver>("Observer2", subject);

	// Attach both observers
	subject->attach(observer1);
	subject->attach(observer2);

	// Detach observer1
	subject->detach(observer1);

	// Trigger an update
	resetOutputStream();
	subject->setState("State3");

	// Assert only observer2 receives the update
	EXPECT_EQ(outputStream.str(), "Observer Observer2 received update.\n");
}

/**
 * @brief Test state synchronization between subject and observers.
 */
TEST(ObserverPatternTest, StateSynchronization) {
	auto subject = std::make_shared<ConcreteSubject>();
	auto observer = std::make_shared<TestObserver>("Observer1", subject);

	// Attach the observer
	subject->attach(observer);

	// Change state in the subject
	subject->setState("SynchronizedState");

	// Verify that the subject's state is correct
	EXPECT_EQ(subject->getState(), "SynchronizedState");
}

/**
 * @brief Test notifying observers without attaching any observers.
 */
TEST(ObserverPatternTest, NotifyWithNoObservers) {
	auto subject = std::make_shared<ConcreteSubject>();

	// Trigger an update without attaching any observers
	resetOutputStream();
	subject->setState("NoObserversState");

	// Assert that no output is produced
	EXPECT_EQ(outputStream.str(), "");
}

File: Behavioral/Observer/python/test_observer.py
import logging
from observer import Subject, ConcreteObserver

# Configure pytest logs
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("observer")


def test_attach_and_notify_observers(caplog):
	"""
	Test attaching multiple observers and verifying notifications.

	Tests that attaching multiple observers and changing the subject state
	result in all observers being notified of the change.
	"""
	subject = Subject()
	observer1 = ConcreteObserver("Observer1")
	observer2 = ConcreteObserver("Observer2")

	# Attach observers
	subject.attach(observer1)
	subject.attach(observer2)

	# Set state and notify
	with caplog.at_level(logging.INFO):
		subject.set_state("State1")

		assert "Subject state updated to: State1" in caplog.text
		assert "Observer Observer1 received update. New state: State1" in caplog.text
		assert "Observer Observer2 received update. New state: State1" in caplog.text


def test_detach_observer(caplog):
	"""
	Test detaching an observer and verifying it no longer receives updates.

	Ensures that after detaching an observer, it does not receive notifications
	when the subject's state changes, while other attached observers do.
	"""

	subject = Subject()
	observer1 = ConcreteObserver("Observer1")
	observer2 = ConcreteObserver("Observer2")

	# Attach and detach observers
	subject.attach(observer1)
	subject.attach(observer2)
	subject.detach(observer2)

	# Set state and notify
	with caplog.at_level(logging.INFO):
		subject.set_state("State2")

		assert "Observer Observer1 received update. New state: State2" in caplog.text
		assert "Observer Observer2 received update. New state: State2" not in caplog.text


def test_no_observers(caplog):
	"""
	Test that the system behaves correctly when no observers are attached.
	
	Ensures that when the subject's state is changed without any observers 
	attached, no observer notifications occur, and only the subject's state 
	update is logged.
	"""

	subject = Subject()

	# Set state without any observers
	with caplog.at_level(logging.INFO):
		subject.set_state("State3")

		assert "Subject state updated to: State3" in caplog.text
		assert "Observer" not in caplog.text


def test_attach_same_observer_multiple_times(caplog):
	"""
	Test attaching the same observer multiple times.

	Verifies that attaching the same observer multiple times results in only
	one notification when the subject's state changes.
	"""
	subject = Subject()
	observer1 = ConcreteObserver("Observer1")

	# Attach the same observer multiple times
	subject.attach(observer1)
	subject.attach(observer1)

	# Set state and notify
	with caplog.at_level(logging.INFO):
		subject.set_state("State4")

		assert caplog.text.count("Observer Observer1 received update. New state: State4") == 1


def test_state_retrieval():
	"""
	Test subject state retrieval after update.
	
	Verifies that the subject's state is updated correctly and can be retrieved
	using the get_state method.
	"""
	subject = Subject()
	subject.set_state("State5")
	assert subject.get_state() == "State5"
